graph-subscriptions-api/datasource/src/datasource_redis.rs

```diff
@@ -0,0 +1,324 @@
+use std::sync::Arc;
+
+use anyhow::{Ok, Result};
+use async_trait::async_trait;
+use chrono::Utc;
+use futures::TryStreamExt;
+use rdkafka::consumer::DefaultConsumerContext;
+use rdkafka::error::KafkaError;
+use rdkafka::{consumer::StreamConsumer, Message};
+use redis::JsonAsyncCommands as _;
+use serde_json::json;
+use tokio::sync::Mutex;
+use toolshed::bytes::{Address, DeploymentId};
+
+use crate::datasource::{Datasource, DatasourceWriter};
+use crate::models::*;
+
+/// The redis datasource implements both the `Datasource` and `DatasourceWriter` traits.
+/// Allows a user to query and store `GatewaySubscriptionQueryResult` records stored in the redis datasource instance.
+pub struct DatasourceRedis {
+    pub redis_client: redis::Client,
+    pub graph_subscriptions_query_result_key: String,
+}
+
+impl DatasourceRedis {
+    pub fn create(addr: &str) -> &'static Self {
+        let client =
+            redis::Client::open(addr).expect("Failure establishing redis client connection");
+
+        Box::leak(Box::new(Self {
+            redis_client: client,
+            graph_subscriptions_query_result_key: String::from(
+                "GATEWAY_SUBSCRIPTION_QUERY_RESULTS",
+            ),
+        }))
+    }
+}
+
+#[async_trait]
+impl Datasource for DatasourceRedis {
+    /// Retrieve the user's unique `RequestTicket` records derived from the stored query result records from the redis database.
+    ///
+    /// # Arguments
+    ///
+    /// - `user` - the user wallet address who performed the stored queries
+    /// - `first` - [OPTIONAL:default 100] the number of records, after sorting, to return
+    /// - `skip` - [OPTIONAL:default 0] the number of records, after sorting, to skip
+    /// - `order_by` - [OPTIONAL] what field on the `RequestTicket` to sort by
+    /// - `order_direction` [OPTIONAL] the sort direction
+    async fn request_tickets(
+        &self,
+        user: Address,
+        first: Option<i32>,
+        skip: Option<i32>,
+        order_by: Option<RequestTicketOrderBy>,
+        order_direction: Option<OrderDirection>,
+    ) -> Result<Vec<RequestTicket>> {
+        let mut conn = self.redis_client.get_async_connection().await?;
+
+        let path = format!("$.{}.*", user.to_string().to_lowercase());
+        let results: Vec<GraphSubscriptionQueryResultRecord> = conn
+            .json_get(&self.graph_subscriptions_query_result_key, path)
+            .await?;
+        if results.is_empty() {
+            return Ok(vec![]);
+        }
+        let mut tickets = RequestTicket::build_unique_request_ticket_list(results);
+        if order_by.is_some() {
+            let direction = order_direction.unwrap_or(OrderDirection::Asc);
+            let order = order_by.unwrap();
```

Comment:
We can pattern match on `order_by` to extract the inner value here to avoid the unwrap
 ```suggestion
        if let Some(order) = order_by {
            let direction = order_direction.unwrap_or(OrderDirection::Asc);
```

Reply:
There are a few more instances of this in this file.